packages/connection/src/Manager.php

```diff
@@ -1370,10 +1370,99 @@ public function verify_secrets( $action, $secret_1, $user_id ) {
 
 	/**
 	 * Responds to a WordPress.com call to authorize the current user.
-	 * Should be changed to protected.
+	 *
+	 * @param array $request The user authorization request data.
+	 * @return array|\IXR_Error Returns an array containing 'result'=>'authorized' on success.
+	 *                          Returns an IXR_Error on failure.
 	 */
-	public function handle_authorization() {
+	public function handle_authorization( $request ) {
+		$user = isset( $request['state'] ) ? get_user_by( 'id', $request['state'] ) : null;
+
+		/**
+		 * We've begun the remote authorization of the user.
+		 *
+		 * @since 8.0.0
+		 *
+		 * @param \WP_User $user The user object.
+		 */
+		do_action( 'jetpack_remote_authorize_begin', $user );
+
+		$return_error = function( \WP_Error $error ) use ( $user ) {
+			/**
+			 * User authorization has failed.
+			 *
+			 * @since 8.0.0
+			 *
+			 * @param \WP_User  $user The user object.
+			 * @param \WP_Error $error The error object.
+			 */
+			do_action( 'jetpack_remote_authorize_fail', $user, $error );
+
+			return $this->convert_to_ixr_error( $error );
+		};
+
+		foreach ( array( 'secret', 'state', 'redirect_uri', 'code' ) as $required ) {
+			if ( ! isset( $request[ $required ] ) || empty( $request[ $required ] ) ) {
+				return $return_error( new \WP_Error( 'missing_parameter', 'One or more parameters is missing from the request.', 400 ) );
+			}
+		}
+
+		if ( ! $user ) {
+			return $return_error( new \WP_Error( 'user_unknown', 'User not found.', 404 ) );
+		}
+
+		if ( $this->is_active() && $this->is_user_connected( $request['state'] ) ) {
+			return $return_error( new \WP_Error( 'already_connected', 'User already connected.', 400 ) );
+		}
 
+		$verified = $this->verify_secrets( 'authorize', $request['secret'], $request['state'] );
+
+		if ( is_wp_error( $verified ) ) {
+			return $return_error( $verified );
+		}
+
+		wp_set_current_user( $request['state'] );
+
+		$client_server = new \Jetpack_Client_Server();
+		$result        = $client_server->authorize( $request );
+
+		if ( is_wp_error( $result ) ) {
+			return $return_error( $result );
+		}
+
+		/**
+		 * We've successfully authorized the user.
+		 *
+		 * @since 8.0.0
+		 *
+		 * @param \WP_User $user The user object.
+		 */
+		do_action( 'jetpack_remote_authorize_success', $user );
+
+		return array(
+			'result' => $result,
+		);
+	}
+
+	/**
+	 * Converts a WP_Error object to an \IXR_Error object.
+	 *
+	 * @param \WP_Error|\IXR_Error $error A WP_Error object.
+	 * @return \IXR_Error
+	 */
+	private function convert_to_ixr_error( $error ) {
```

Comment:
There are a few other methods in this package that also convert WP_Error objects to IXR_Error objects:
* [XMLRPC_Connector](https://github.com/Automattic/jetpack/blob/master/packages/connection/src/XMLRPC_Connector.php#L65)
* [jetpack-xmlrpc-server](https://github.com/Automattic/jetpack/blob/master/packages/connection/legacy/class.jetpack-xmlrpc-server.php#L626)

There's been some [discussion](https://github.com/Automattic/jetpack/pull/13444#discussion_r322223042) about adding a Utils class to this package. When the Utils class is added, we should consolidate all of these methods and put it into the Utils class.